### ISSUE TITLE
UsageTracker: track the least-recently-used blobs in vineyard server.

### ISSUE THREAD

**Issue by septicmk** [v6d-io/v6d#739]
As titled.
- [x] #765 
- [ ] (TODO) Apply LRU on cold blobs.

**Comment by ZjuYTW** [comment]
I am about to add this feature. And the following is a draft-plan, appreciate for any advice!

First achieve a singleton LRU class
```
class LRU {
    using lru_iter = std::list<ObjectID>::iterator;
    using clru_iter = std::list<ObjectID>::const_iterator;

    public:
    static LRU* GetInstance();

    void Refer(ObjectID obj_id) const;

    clru_iter begin() const; // used for iterating through the list

    private:
    LRU() {} // NOLINT

    std::list<ObjectID> cache_;
    std::unordered_map<ObjectID, lru_iter> iter_table_;
    size_t max_size_; // Maybe we need a threshold to keep track of those hot blob

    static LRU* lru_;
  }
```

Then we should call Lru::refer every time we try to visit blob
```
const std::shared_ptr<arrow::Buffer>& Blob::Buffer() const {
  ...
  Lru::GetInstance()->refer(this->id_);
  return buffer_;
}
```

A obvious drawback is that we can't refresh the lru list everytime the buffer_ is used...

**Comment by sighingnow** [comment]
Ke @septicmk  may you give some comment on the proposal above?

Thanks!

**Comment by septicmk** [comment]
Hi @ZjuYTW, An LRU cache may not be enough for later spilling. For example, we should not spill an object already in use. I suggest we should track the usage of every object, and only maintain LRU caches for unused objects. This means we should record something like “reference count” and only put the objects that "reference count==0"  into your LRU cache.

FYI, we already have a usage tracker on the client-side to maintain the reference count (https://github.com/v6d-io/v6d/blob/main/src/client/client.h#L122), we could have something similar on the server-side.

**Comment by ZjuYTW** [comment]
@septicmk  Sry for the lag on response (little bit busy in previous weeks). As you mentioned, we need to keep track of all objects in the server side.
And there are two BulkStore type in vineyard_server, and seems we don't need to implement UsageTrack on `BulkStore` ? 
For `PlasmaBulkStore`, because already have a `reference counter` CRTP to tracker reference of object, we may add the LRU cache in the PlasmaBulkStore's as a private member to track the unused objects.

Then whenever the `Release` is triggered and caused ref_cnt becomes 0. Then `PlasmaBulkStore::OnRelease` will be called . We may add LRU::Add(id) to add the deleted ID into the LRU cache here.
Also, when using the object. We should check the LRU and if the ID exists in there, just erase the ID in the LRU cache.
The structure of LRU still looks seem as previous.

In this way, later we could refactor the Memory Free step into the LRU for memory management.

**Comment by septicmk** [comment]
@ZjuYTW Thanks for your idea! The `PlasmaBulkStore` is designed for compatibility. What we want to do here is to track the usage of the original `BulkStore` which is actually used in our production environment (So you can forget the `PlasmaBulkStore` for now). You may notice that both client-side and server-side have CRTP to track the reference counts, this is just an optimization to alleviate IPC overhead. Tracking reference count for blobs is a little bit complex(have to consider deletion, shallow copy, blob slicing, and client-side failure), I will push some commits these days to maintain a list of cold objects, then you can intuitively apply the LRU algorithm on them to pop victims for spilling.

**Comment by sighingnow** [comment]
Closed by #794.
